### PR TITLE
add equality and hashing abilities + convenience functions to columns

### PR DESCRIPTION
**Pull request by tobymao**
so i first tried implementing AST comparison. and that worked, but it was far more lines of code 

i figure it's possible for the AST to differ but the resulting SQL be the same due to optimizations, this actually happens when we parse IF vs CASE, the AST is different, but the resulting SQL is the same due to optimization in the Generator.

so anyways, I choose to have the identified non-pretty sql be the source of truth

### CODE REVIEW

**Comment by izeigerman** [comment]
**TL; DR**
I don't think we can account for all possible ways to optimize the tree, so we shouldn't even try to incorporate optimizations into the tree comparison logic.

**Longer version**

> i figure it's possible for the AST to differ but the resulting SQL be the same due to optimizations, this actually happens when we parse IF vs CASE, the AST is different, but the resulting SQL is the same due to optimization in the Generator.

I totally understand the motivation here, but I, as a user, would prefer comparison of actual nodes and returning false in the IF vs CASE scenario makes more sense to me. I do believe the tree comparison should be completely independent from either Parser or Generator.

My primary concern arises from use cases with a logic which involves tree traversal changes based on a result of the equality comparison. If the comparison returns true while actual tree structures are different this may lead to unexpected behaviors. 

Please also keep in mind that interpreting a tree as a SQL string may potentially be only one way of interpretation and more use cases may arise which utilize the AST in a completely different way. One example of this may be external optimizers, which also means that all optimization scenario can't possibly be taken into account during comparison. 

I also don't think that the following statement is especially consistent:
> but the resulting SQL is the same due to optimization in the Generator.

Provided we always use the default Dialect for comparison, we completely ignore the fact that some of these optimizations can be introduced by specific dialects, which again means that the initial assumption is not complete.

File: sqlglot/expressions.py [hunk 1]
```diff
@@ -167,7 +175,23 @@ class CTE(Expression):
 
 class Column(Expression):
     token_type = TokenType.COLUMN
-    arg_types = {"this": True, "db": False, "table": False, "fields": False}
+    arg_types = {"this": False, "db": False, "table": False, "fields": False}
```
Comment (izeigerman) [inline sqlglot/expressions.py:178]: Why is `this` no longer required?

File: sqlglot/expressions.py [hunk 2]
```diff
@@ -167,7 +175,23 @@ class CTE(Expression):
...
+
+    def text(self, kind):
+        arg = self.args.get(kind)
+        return arg.text if arg else ""
```
Comment (izeigerman) [inline sqlglot/expressions.py:182]: [Optional/Nitpick] Shall we ensure that it's actually a token?

File: sqlglot/tokens.py [hunk 3]
```diff
@@ -152,6 +152,7 @@ class TokenType(AutoName):
     UPDATE = auto()
     VALUES = auto()
     VIEW = auto()
+    WINDOW = auto()
```
Comment (izeigerman) [inline sqlglot/tokens.py:155]: It seems like I don't fully understand the logic behind the tokenizer. I see that we added a new token type but I don't see a mapping for it in the Tokenizer's `KEYWORDS`.

File: sqlglot/expressions.py (hunk 1 above)
Comment (tobymao) [reply sqlglot/expressions.py:178]: because if we set fields (a.b.c.d.e) i make it null, basically highly nested structs

File: sqlglot/expressions.py (hunk 2 above)
Comment (tobymao) [reply sqlglot/expressions.py:182]: i think it should always be a token? if not it's a bug

File: sqlglot/tokens.py (hunk 3 above)
Comment (tobymao) [reply sqlglot/tokens.py:155]: it's to make expression and token duck typed

Comment (tobymao) [reply sqlglot/tokens.py:155]: because expression needs to have a corresponding token_type

Comment (tobymao) [reply sqlglot/tokens.py:155]: maybe this should be changed?

File: sqlglot/expressions.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/expressions.py:178]: ah, that makes sense. I wonder whether we need some sort of a custom validation here. Like eg either this or that argument must be present.

File: sqlglot/tokens.py (hunk 3 above)
Comment (izeigerman) [reply sqlglot/tokens.py:155]: Not a blocker for this PR but perhaps worth revisiting at some point.

File: sqlglot/expressions.py (hunk 2 above)
Comment (izeigerman) [reply sqlglot/expressions.py:182]: Yeah, I was just wondering whether we wanted a more specific error message in this as oppose to generic attribute not found error.

File: sqlglot/expressions.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/expressions.py:178]: btw, parsing of deeply nested structs looks weird but doesn't fail with the current implementation:
```python
>>> from sqlglot import parse_one
>>> parse_one("a.b")
(COLUMN this: b, db: , table: a, fields: )
>>> parse_one("a.b.c")
(COLUMN this: c, db: a, table: b, fields: )
>>> parse_one("a.b.c.d")
(COLUMN this: c, db: a, table: b, fields: a, b, c, d)
>>> parse_one("a.b.c.d.e")
(COLUMN this: c, db: a, table: b, fields: a, b, c, d, e)
>>> parse_one("a.b.c.d.e.f")
(COLUMN this: c, db: a, table: b, fields: a, b, c, d, e, f)
```

**Review by izeigerman — APPROVED** [review]
